### PR TITLE
miri: add write_bytes method to Memory doing bounds-checks and supporting iterators

### PR DESCRIPTION
**Pull request by RalfJung**
This lets us avoid some direct `Allocation` accesses in Miri.

### CODE REVIEW

File: src/librustc/mir/interpret/allocation.rs [hunk 1]
```diff
@@ -346,11 +346,17 @@ impl<'tcx, Tag: Copy, Extra: AllocationExtra<Tag>> Allocation<Tag, Extra> {
         &mut self,
         cx: &impl HasDataLayout,
         ptr: Pointer<Tag>,
-        src: &[u8],
+        src: impl IntoIterator<Item=u8, IntoIter: iter::ExactSizeIterator>,
     ) -> InterpResult<'tcx>
```
Comment (christianpoveda) [inline src/librustc/mir/interpret/allocation.rs]: Maybe there should be a comment suggesting to use `Memory::write_bytes` instead? I think that unless someone is writing in several calls, they should use the other function

Comment (RalfJung) [reply src/librustc/mir/interpret/allocation.rs]: Well that applies to every single function in this module...

Comment (christianpoveda) [reply src/librustc/mir/interpret/allocation.rs]: Ok that's true

**Pull request merged**